### PR TITLE
Tinymce customization

### PR DESCRIPTION
**Pull request by stevepiercy**
Closes #1806.

----
📚 Documentation preview 📚: https://plone6--1807.org.readthedocs.build/

### CODE REVIEW

File: docs/classic-ui/tinymce-customization.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+# TinyMCE customization
+
+This chapter is a developer reference manual for customizing {term}`TinyMCE`.
+
+
+(tinymce-remove-formats-label)=
+
+## Remove formats
+
+In Plone 6, TinyMCE has a special logic that automatically reads registered files named {file}`tinymce-formats.css` and adds the CSS classes defined in those files to TinyMCE's {menuselection}`Format --> Formats` menu.
```
Comment (petschki) [inline docs/classic-ui/tinymce-customization.md]:
```suggestion
In Plone 6, TinyMCE has a special logic that automatically reads registered files named {file}`tinymce-formats.css` and adds the CSS classes defined in those files to TinyMCE's {menuselection}`Format --> Formats` menu by using the [`importcss_file_filter` option](https://www.tiny.cloud/docs/tinymce/latest/importcss/#importcss_file_filter).
```

**Review by petschki — CHANGES_REQUESTED** [review]

**Comment by stevepiercy** [comment]
Direct link to preview https://plone6--1807.org.readthedocs.build/classic-ui/tinymce-customization.html

**Comment by petschki** [comment]
@stevepiercy I've changed the order of information and enhanced it:

1. customize formats (with recommended JSON config option and example how to update)
2. remove formats

**Comment by stevepiercy** [comment]
@petschki I made a few more revisions to better explain why someone should choose one option over another, and turned the .css file option into more of a how-to-guide.

**Review by petschki — APPROVED** [review]
👍🏼

**Comment by szakitibi** [comment]
Please note following the instructions of [Edit the Formats option in the TinyMCE control panel](https://plone6--1807.org.readthedocs.build/classic-ui/tinymce-customization.html#edit-the-formats-option-in-the-tinymce-control-panel) will not add the new item to the "Format" menu. Editing the "Formats" JSON will only add the format to the configurations, but it will be still invisible in the TinyMCE's menu. E.g. "discreet" for classic demo portal is registered as one of the "Formats" in the JSON, but it is nowhere to be found in the TinyMCE menu.

As a last step "Formats" must be registered under one of styles (in the case of "myformat", under "Block styles" of the controlpanel), only then will it pop up in the TinyMCE's menu.

File: docs/classic-ui/tinymce-customization.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+They are less complicated to implement than the third option.
+
+
+### Add-on GenericSetup configuration file
+
+This option is best for system administrators and developers who write their own add-ons to ease reproducibility.
+
+You can add a GenericSetup configuration file to your add-on, such as {file}`profiles/default/registry/tinymce.xml`, with the following content.
+
+```xml
```
Comment (yurj) [inline docs/classic-ui/tinymce-customization.md:31]: What about
```xml
  <records interface="Products.CMFPlone.interfaces.ITinyMCESchema" prefix="plone">
    <value key="other_settings">
      {
        "extended_valid_elements":"i[class],script[type|src]",
        "style_formats": [
          {
            "title": "Highlight",
            "block": "p",
            "attributes": {
              "class": "featured"
            }
          },
          {
            "title": "Alert",
            "block": "p",
            "attributes": {
              "class": "redalert"
            }
          }
        ],
        "style_formats_merge": true
      }
    </value>
  </records>
```

File: docs/classic-ui/tinymce-customization.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+1.  Name the file {file}`tinymce-formats.css`.
+1.  Register the file as a resource in your Plone site.
+1.  Include the file in the TinyMCE control panel in the textarea input {guilabel}`Choose the CSS used in WYSIWYG Editor Area`.
+    1.  Navigate to {menuselection}`Site Setup --> TinyMCE --> Default`, or append `@@tinymce-controlpanel` to the root of your website in your browser's location bar.
+    1.  Scroll down to {guilabel}`Choose the CSS used in WYSIWYG Editor Area`.
+
+CSS styles defined in this file will automatically be added to the top level of TinyMCE's {menuselection}`Format --> Formats` menu.
+
+
+## Remove imported formats
```
Comment (yurj) [inline docs/classic-ui/tinymce-customization.md:77]: is there a generic setup xml example to do this? This would pair with the above example of a GS xml.

**Comment by petschki** [comment]
@szakitibi yes you're right. forgot about that. But depending on the format you can also add it to the "Inline formats"

File: docs/classic-ui/tinymce-customization.md (hunk 3 above)
Comment (petschki) [reply docs/classic-ui/tinymce-customization.md:77]: you mean removing it? sure... just add the updated record to the xml and add `purge="true"` as parameter.

```
<record field="content_css"
          interface="plone.base.interfaces.controlpanel.ITinyMCESchema"
          name="plone.content_css"
  >
    <value purge="true">
      <element>++theme++barceloneta/tinymce/tinymce-ui-content.css</element>
    </value>
  </record>
```

feel free to open a PR and enhance it ...

**Comment by szakitibi** [comment]
> @szakitibi yes you're right. forgot about that. But depending on the format you can also add it to the "Inline formats"

Exactly. That is why I used "under one of the styles". The official documentation should be clear about this part, that the registered "formats" can be registered under different styles, and that will determine their position in the TinyMCE menu.

**Comment by petschki** [comment]
Last but not least there should be a warning, that the provided genericsetup xml will kill the default formats. If those should be preserved, you need to copy them over to your `tinymce.xml`

File: docs/classic-ui/tinymce-customization.md (hunk 2 above)
Comment (petschki) [reply docs/classic-ui/tinymce-customization.md:31]: Beware: `other_settings` will override all previously made settings. This is meant to inject settings which are not populated in other config fields.

And the `style_formats_merge` options provides the default "tinymce styles" but that's not exactly the same as the plone default tinymce styles ...

**Comment by szakitibi** [comment]
Also another comment regarding the use of "formats" JSON. I think it is not possible to add top level items (like the `highlight-inline`) to "Formats" dropdown menu with the available "* styles" configurations. It seems currently only the `tinymce-formats.css` is able to do that trick.

**Comment by stevepiercy** [comment]
I would be happy to accept a pull request with revisions to this document. These comments are great, but they won't get into the documentation unless someone who has the technical knowledge writes them up in PR.

I do not have this technical knowledge. I am only a technical writer who translates Plone-ish to English and sprinkles some MyST syntax on top to make it pretty. Thanks for your collaboration.

File: docs/classic-ui/tinymce-customization.md (hunk 3 above)
Comment (stevepiercy) [reply docs/classic-ui/tinymce-customization.md:77]: Done in 43807fa17a507b407dcc353ec812461eeadccbf9

**Pull request merged**